dummyfs/dummyfs.c

```diff
@@ -28,470 +29,597 @@
 
 	if (err < 0) {
+		mutexUnlock(fs->mutex);
 		return err;
 	}
 
-	o = dummyfs_get(fs, res);
-	object_lock(fs, o);
+	o = dummyfs_object_get(fs, res);
 	*dev = o->dev;
-	object_unlock(fs, o);
-	object_put(fs, o);
+	dummyfs_object_put(fs, o);
+	mutexUnlock(fs->mutex);
 
 	return len;
 }
 
 
-int dummyfs_setattr(void *ctx, oid_t *oid, int type, long long attr, const void *data, size_t size)
+static int _dummyfs_truncateObject(dummyfs_t *fs, dummyfs_object_t *o, size_t size)
 {
-	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
-	int ret = EOK;
+	TRACE;
+	/* Lazy: start allocating memory only after first write */
+	if ((size == o->size) || (o->data == NULL) || (o->chunks == NULL)) {
+		o->size = size;
+		return 0;
+	}
 
-	if ((o = dummyfs_get(fs, oid)) == NULL)
-		return -ENOENT;
+	if (o->size < DUMMYFS_CHUNKSZ) {
+		/* Small file */
+		if (size >= DUMMYFS_CHUNKSZ) {
+			/* Make a big file */
+			void *page = dummyfs_mmap(fs);
+			if (page == NULL) {
+				return -ENOMEM;
+			}
 
-	object_lock(fs, o);
-	switch (type) {
-		case (atUid):
-			o->uid = attr;
-			break;
+			void **chunks = dummyfs_calloc(fs, sizeof(void *) * DUMMYFS_CHUNKCNT(size));
+			if (chunks == NULL) {
+				dummyfs_munmap(fs, page);
+				return -ENOMEM;
+			}
 
-		case (atGid):
-			o->gid = attr;
-			break;
+			memcpy(page, o->data, o->size);
 
-		case (atMode):
-			o->mode = (o->mode & ~ALLPERMS) | (attr & ALLPERMS);
-			break;
+			dummyfs_free(fs, o->data, o->size);
 
-		case (atSize):
-			object_unlock(fs, o);
-			ret = dummyfs_truncate(ctx, oid, attr);
-			object_lock(fs, o);
-			break;
+			chunks[0] = page;
+			o->chunks = chunks;
+		}
+		else {
+			void *tptr = dummyfs_realloc(fs, o->data, o->size, size);
+			if ((size != 0) && (tptr == NULL)) {
+				return -ENOMEM;
+			}
 
-		case (atPort):
-			ret = -EINVAL;
-			break;
+			o->data = tptr;
 
-		case (atDev):
-			if (data != NULL && size == sizeof(oid_t) && S_ISDIR(o->mode) && dummyfs_device(fs, &o->dev) == 0) {
-				object_unlock(fs, o);
-				ret = dev_create(fs, (oid_t *)data, o);
-				object_lock(fs, o);
+			if ((size != 0) && (size > o->size)) {
+				memset((char *)o->data + o->size, 0, size - o->size);
 			}
-			else {
-				ret = -EINVAL;
+		}
+	}
+	else {
+		/* Big files */
+		if (size < DUMMYFS_CHUNKSZ) {
+			/* Make small file */
+			void *tptr = NULL;
+
+			if (size != 0) {
+				tptr = dummyfs_malloc(fs, size);
+				if (tptr == NULL) {
+					return -ENOMEM;
+				}
+
+				memcpy(tptr, o->chunks[0], size);
 			}
-			break;
 
-		case (atMTime):
-			o->mtime = attr;
-			break;
+			size_t limit = DUMMYFS_CHUNKCNT(o->size);
+			for (size_t i = 0; i < limit; ++i) {
+				if (o->chunks[i] != NULL) {
+					dummyfs_munmap(fs, o->chunks[i]);
+				}
+			}
 
-		case (atATime):
-			o->atime = attr;
-			break;
+			dummyfs_free(fs, o->chunks, sizeof(void *) * DUMMYFS_CHUNKCNT(o->size));
 
-		default:
-			ret = -EINVAL;
-			break;
+			o->data = tptr;
+		}
+		else {
+			if (DUMMYFS_CHUNKCNT(size) != DUMMYFS_CHUNKCNT(o->size)) {
```

Comment:
No point in doing that - we ignore rest of the chunk in the dummyfs_read explicitly.

This way I would rather need to be concerned about going out of buffer in case of using o->data

Reply:
I was thinking about the case of user storing eg password in dummy fs then truncating the file to remove it. Then  it would be best if the password was immediatelly removed. However, I am not sure about real  life implications and validity of even stroing password in fs

Reply:
There's no way to access this data anyway, apart from accessing dummyfs memory map directly